tests/frontier/scenarios/common.py

```diff
@@ -0,0 +1,302 @@
+) -> Bytecode:
+    """
+    Run find replace of some special calls to the contracts that we don't know at compile time
+    replace 0xfff..fff address to external_address
+    replace special call to 0xfff..ffe address to gas_hash_address contract.
+    """
+    gas_hash_address = make_gas_hash_contract(pre)
+    invalid_opcode_contract = make_invalid_opcode_contract(pre, fork)
+
+    replace_list: List[Tuple[str, str]] = [
+        (
+            "602060646020604060007f"
+            + hex(SpecialAddress.INVALID_CALL_ADDRESS)[2:].lower()
+            + "611388f1",
+            Op.CALL(Op.SUB(Op.GAS, 200000), invalid_opcode_contract, 0, 64, 32, 100, 32).hex(),
+        ),
+        (
+            "600060006020604060007f" + hex(SpecialAddress.GAS_HASH_ADDRESS)[2:].lower() + "5af1",
+            Op.CALL(Op.SUB(Op.GAS, 200000), gas_hash_address, 0, 64, 32, 0, 0).hex(),
+        ),
+        (
+            "7f" + hex(SpecialAddress.EXTERNAL_ADDRESS)[2:].lower() + "31",
+            Op.BALANCE(external_address).hex(),
+        ),
+        (
+            "7f" + hex(SpecialAddress.EXTERNAL_ADDRESS)[2:].lower() + "3b",
+            Op.EXTCODESIZE(external_address).hex(),
+        ),
+        (
+            "7f" + hex(SpecialAddress.EXTERNAL_ADDRESS)[2:].lower() + "3f",
+            Op.EXTCODEHASH(external_address).hex(),
+        ),
+        (
+            "7f" + hex(SpecialAddress.EXTERNAL_ADDRESS)[2:].lower() + "3c",
+            "3c".join(Op.BALANCE(external_address).hex().rsplit("31", 1)),
+        ),
+    ]
+    for find, replace in replace_list:
+        if find in operation.hex():
+            new_operation_hex = operation.hex().replace(
+                find,
+                replace,
+            )
+            operation = Bytecode(
+                bytes_or_byte_code_base=to_bytes(new_operation_hex),
+                popped_stack_items=0,
+                pushed_stack_items=0,
+            )
+    return operation
+
+
+def make_gas_hash_contract(pre: Alloc) -> Address:
+    """
+    Contract that spends unique amount of gas based on input
+    Used for the values we can't predict, can be gas consuming on high values
+    So that if we can't check exact value in expect section,
+    we at least could spend unique gas amount.
+    """
+    gas_hash_address = pre.deploy_contract(
+        code=Op.MSTORE(0, 0)
+        + Op.JUMPDEST
+        + Op.CALLDATACOPY(63, Op.MLOAD(0), 1)
+        + Op.JUMPDEST
+        + Conditional(
+            condition=Op.EQ(Op.MLOAD(32), 0),
+            if_true=Op.MSTORE(0, Op.ADD(1, Op.MLOAD(0)))
+            + Conditional(
+                condition=Op.GT(Op.MLOAD(0), 32),
+                if_true=Op.RETURN(0, 0),
+                if_false=Op.JUMP(5),
+            ),
+            if_false=Op.MSTORE(32, Op.SUB(Op.MLOAD(32), 1)) + Op.JUMP(14),
+        )
+    )
+    return gas_hash_address
+
+
+def make_invalid_opcode_contract(pre: Alloc, fork: Fork) -> Address:
```

Comment:
I think we can make this its own standalone test in a different .py file and just parametrize a test function with all invalid opcodes, instead merging everything into a single bytecode.